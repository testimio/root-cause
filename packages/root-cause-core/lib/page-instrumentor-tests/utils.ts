import type { RootCausePage } from '../interfaces';
import { PuppeteerPageHooker } from '../PuppeteerPageHooker';
import { TestContextMemory } from '../TestContextMemory';
import { createDateMocker } from '../utils';

export async function runPageInstrumentorTest<TPage extends RootCausePage>(
  testName: string,
  fullTestFileName: string,
  page: TPage,
  testCode: (wrappedPage: TPage) => Promise<void>
): Promise<SetupPageInstrumentorForTestingReturn<TPage>['hooks']> {
  const { pageInstrumentor, wrappedPage, hooks } = await setupPageInstrumentorForTesting(
    testName,
    fullTestFileName,
    page
  );

  pageInstrumentor.start();

  await testCode(wrappedPage);

  await pageInstrumentor.end({
    success: true,
  });

  return hooks;
}

interface SetupPageInstrumentorForTestingReturn<TPage extends RootCausePage> {
  wrappedPage: TPage;
  pageInstrumentor: PuppeteerPageHooker;
  hooks: {
    beforeAllHook: jest.Mock<any, any>;
    afterAllHook: jest.Mock;
    beforeEachHook: jest.Mock;
    afterEachHook: jest.Mock;
  };
}

export async function setupPageInstrumentorForTesting<TPage extends RootCausePage>(
  testName: string,
  fullTestFileName: string,
  page: TPage
): Promise<SetupPageInstrumentorForTestingReturn<TPage>> {
  const testContext = new TestContextMemory(
    '/tmp',
    testName,
    `full ${testName}`,
    fullTestFileName,
    {
      screenshots: false,
      console: false,
      networkLogs: false,
      jestAssertions: false,
      html: false,
    },
    createDateMocker()
  );

  const pageInstrumentor = new PuppeteerPageHooker(testContext, page);

  const wrappedPage = pageInstrumentor.wrapWithProxy(page);

  const beforeAllHook = jest.fn();
  const afterAllHook = jest.fn();

  const beforeEachHook = jest.fn();
  const afterEachHook = jest.fn();

  pageInstrumentor.registerBeforeAllHook(beforeAllHook);
  pageInstrumentor.registerAfterAllHook(afterAllHook);
  pageInstrumentor.registerBeforeHook(beforeEachHook);
  pageInstrumentor.registerAfterHook(afterEachHook);

  return {
    wrappedPage,
    pageInstrumentor,
    hooks: {
      beforeAllHook,
      afterAllHook,
      beforeEachHook,
      afterEachHook,
    },
  };
}

const objPropToClean = ['testContext', 'proxyContext', 'rootPage', 'instrumentedFunctionResult'];
export function cleanJestMockFunctionCalls(calls: any[]): any[] {
  return calls.map((callArg) => {
    const cloned = { ...callArg };

    for (const prop of objPropToClean) {
      if (prop in cloned) {
        cloned[prop] = 'cleaned';
      }
    }

    return cloned;
  });
}
