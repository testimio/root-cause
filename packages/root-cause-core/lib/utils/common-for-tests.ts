export const fakeOfType = (name: string) => ({ constructor: { name } });

export const Page = fakeOfType('Page');
export const Keyboard = fakeOfType('Keyboard');
export const ElementHandle = fakeOfType('ElementHandle');
export const Mouse = fakeOfType('Mouse');
export const Frame = fakeOfType('Frame');
