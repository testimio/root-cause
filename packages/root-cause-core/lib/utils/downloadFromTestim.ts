import { TestimApi } from '../testim-services-api';
import { getTestimCredentials } from '../testim-credentials-manager';

async function authenticateIfNeeded(): Promise<TestimApi | null> {
  const testimAuthInfo = await getTestimCredentials();

  if (testimAuthInfo.projectId && testimAuthInfo.ciToken) {
    const testim = new TestimApi();
    await testim.authenticate(testimAuthInfo.projectId, testimAuthInfo.ciToken);
    return testim;
  }
  return null;
}
const testimBackendPromise = authenticateIfNeeded().catch(() => {});

export async function getCloudAsset(path: string) {
  const testim = await testimBackendPromise;
  if (!testim || !testim.projectId) {
    throw new Error('Unauthenticated');
  }

  const response = await testim.assets.getAsset({
    projectId: testim.projectId,
    pathInsideBucket: `.root-cause/${path}`,
  });
  return response;
}
