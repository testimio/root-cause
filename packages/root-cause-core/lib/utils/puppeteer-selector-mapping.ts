import { Page } from 'puppeteer';

const firstArg = (args: any[]) => args[0];
type SelectorMapping = { [P in keyof Page]?: (args: any[]) => string | undefined };

export const extractPuppeteerSelectorMapping = ({
  $: firstArg,
  $$: firstArg,
  $$eval: firstArg,
  $eval: firstArg,
  $x: Function.prototype, // TODO(Benji) supporting xpath here should be pretty easy
  focus: firstArg,
  hover: firstArg,
  click: firstArg,
  type: firstArg,
  tap: firstArg,
  select: firstArg,
  waitFor(args: any[]) {
    if (typeof args[0] === 'string') {
      return args[0];
    }
    return undefined;
  },
  waitForSelector: firstArg,
} as const) as SelectorMapping;

export function extractPuppeteerSelector(fnName: keyof Page, args: any[]): string | undefined {
  const extractor = extractPuppeteerSelectorMapping[fnName] || (() => undefined);
  return extractor(args);
}
