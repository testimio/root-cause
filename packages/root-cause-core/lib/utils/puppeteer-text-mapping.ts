import { RootCausePage } from '../interfaces';

const secondArg = (args: any[]) => args[1];
const firstArg = (args: any[]) => args[0];
const returnValue = (_: any[], returnValue: any) =>
  typeof returnValue === 'string' ? returnValue : undefined;
type ExtractTextMapping = {
  [P in keyof RootCausePage]?: (args: any[], returnValue: any) => string | undefined;
};

export const extractPuppeteerTextMapping = ({
  type: secondArg,
  evaluate: returnValue,
  goto: firstArg,
  $eval: returnValue,
} as const) as ExtractTextMapping;

export function extractPuppeteerText(
  fnName: keyof RootCausePage,
  args: any[],
  returnValue: any
): string | undefined {
  const extractor = extractPuppeteerTextMapping[fnName] || (() => undefined);
  return extractor(args, returnValue);
}
