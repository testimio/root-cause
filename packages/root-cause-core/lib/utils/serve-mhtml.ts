import debug from 'debug';
import type { RequestHandler, Response } from 'express';
import { IFileResult, Parser } from 'fast-mhtml';
import fs from 'fs-extra';
import path from 'path';

const logger = debug('root-cause:debug:serve-mhtml');

function sendFile(res: Response, file: IFileResult) {
  logger('serve-mhtml: sending file %s', file.filename);
  res.setHeader('Content-Type', file.type);
  res.send(file.content);
  res.end();
}

function failWith(res: Response, status: number, error: string) {
  logger('serve-mhtml: %s', error);
  res.status(status);
  res.send(error);
  res.end();
}

export function serveMhtml(testPath: string): RequestHandler {
  const fileCache = new Map<string, IFileResult>();

  async function processMhtml(file: string): Promise<IFileResult> {
    fileCache.clear();

    const data = await fs.readFile(path.resolve(testPath, file));

    const splitFiles: IFileResult[] = new Parser({}).parse(data).rewrite().spit();
    for (const result of splitFiles) {
      fileCache.set(result.filename.replace(/#.*/, ''), result); // remove hash and set in cache
    }

    return splitFiles[0];
  }

  return function serveMhtmlMiddleware({ params: { file } }, res) {
    (async () => {
      if (file.endsWith('mhtml')) {
        const indexFromHtml = await processMhtml(file);
        return sendFile(res, indexFromHtml);
      }

      const result = fileCache.get(file);
      if (!result) {
        return failWith(
          res,
          404,
          `Missing ${file}, should be one of: ${JSON.stringify(fileCache.keys())}`
        );
      }

      sendFile(res, result);
    })().catch((error) => {
      return failWith(res, 500, `Error: ${error}<br />${error.stack.replace(/\n/, '<br />')}`);
    });
  };
}
