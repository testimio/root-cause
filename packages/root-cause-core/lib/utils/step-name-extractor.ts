import type { StepResult } from '@testim/root-cause-types';

export function extractStepName(result: StepResult) {
  const name = result.fnName ?? 'Run Command';
  const selector = result.selector;
  const text = result.text;
  if (selector && text) {
    return `${name} "${text}" on "${selector}"`;
  }
  if (selector) {
    return `${name} "${selector}"`;
  }
  if (text) {
    return `${name} "${text}"`;
  }
  return `${name}`;
}
