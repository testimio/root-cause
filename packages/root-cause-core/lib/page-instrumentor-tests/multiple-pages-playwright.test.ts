import {
  getCleanAllPathsPrettyFormatPlugin,
  getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin,
} from '@testim/internal-self-tests-helpers';
import playwright from 'playwright';
import { cleanJestMockFunctionCalls, runPageInstrumentorTest } from './utils';
import { sleep } from '../utils';

describe('Multiple pages', () => {
  jest.setTimeout(30_000);
  expect.addSnapshotSerializer(getCleanAllPathsPrettyFormatPlugin(process.cwd()));
  expect.addSnapshotSerializer(getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin());

  describe('Playwright chromium', () => {
    let browser: playwright.Browser;
    let browserContext: playwright.BrowserContext;
    let page: playwright.Page;

    beforeAll(async () => {
      // https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#setting-up-chrome-linux-sandbox
      // Not optimal, but didn't work on circle on circleci/node:12.17-stretch-browsers without it
      // Need to revisit
      browser = await playwright.chromium.launch({
        args: ['--no-sandbox', '--disable-setuid-sandbox'],
      });
    });

    afterAll(async () => {
      await browser.close();
    });

    beforeEach(async () => {
      browserContext = await browser.newContext();
      page = await browserContext.newPage();
    });

    afterEach(async () => {
      await page.close();
      await browserContext.close();
    });

    test('page.once, ensure page.off is working', async () => {
      await runPageInstrumentorTest('page.once method', __filename, page, async (wrappedPage) => {
        await wrappedPage.setContent('<a href="about:blank" target="blank">CLICK HERE</a>');

        const onceHandler = jest.fn();
        wrappedPage.once('popup', onceHandler);
        wrappedPage.off('popup', onceHandler);

        await wrappedPage.click('a');

        // wait some arbitrary time
        await sleep(100);
        expect(onceHandler).not.toBeCalled();
      });
    });

    test('page.once, ensure page.removeListener is working', async () => {
      await runPageInstrumentorTest('page.once method', __filename, page, async (wrappedPage) => {
        await wrappedPage.setContent('<a href="about:blank" target="blank">CLICK HERE</a>');

        const onceHandler = jest.fn();
        wrappedPage.once('popup', onceHandler);
        wrappedPage.removeListener('popup', onceHandler);

        await wrappedPage.click('a');

        // wait some arbitrary time
        await sleep(100);
        expect(onceHandler).not.toBeCalled();
      });
    });

    test('page.addListener, ensure page.removeListener is working', async () => {
      await runPageInstrumentorTest('page.once method', __filename, page, async (wrappedPage) => {
        await wrappedPage.setContent('<a href="about:blank" target="blank">CLICK HERE</a>');

        const onceHandler = jest.fn();
        wrappedPage.addListener('popup', onceHandler);
        wrappedPage.removeListener('popup', onceHandler);

        await wrappedPage.click('a');

        // wait some arbitrary time
        await sleep(100);
        expect(onceHandler).not.toBeCalled();
      });
    });

    test('page.on, ensure page.off is working', async () => {
      await runPageInstrumentorTest('page.once method', __filename, page, async (wrappedPage) => {
        await wrappedPage.setContent('<a href="about:blank" target="blank">CLICK HERE</a>');

        const onceHandler = jest.fn();
        wrappedPage.on('popup', onceHandler);
        wrappedPage.off('popup', onceHandler);

        await wrappedPage.click('a');

        // wait some arbitrary time
        await sleep(100);
        expect(onceHandler).not.toBeCalled();
      });
    });

    test('page.addListener("popup") method', async () => {
      const {
        beforeAllHook,
        afterAllHook,
        beforeEachHook,
        afterEachHook,
      } = await runPageInstrumentorTest(
        'page.addListener("popup") method',
        __filename,
        page,
        async (wrappedPage) => {
          await wrappedPage.setContent('<a href="about:blank" target="blank">CLICK HERE</a>');

          const [, newPage] = await Promise.all([
            wrappedPage.click('a'),
            new Promise<playwright.Page>((res) => {
              wrappedPage.addListener('popup', res);
            }),
          ]);
          await newPage.click('body');
        }
      );

      expect(beforeAllHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchSnapshot(
        'beforeAllHook'
      );
      expect(afterAllHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchSnapshot(
        'afterAllHook'
      );

      expect(beforeEachHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchSnapshot(
        'beforeEachHook'
      );

      expect(afterEachHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchSnapshot(
        'afterEachHook'
      );
    });

    test('page.on("popup") method', async () => {
      const {
        beforeAllHook,
        afterAllHook,
        beforeEachHook,
        afterEachHook,
      } = await runPageInstrumentorTest(
        'page.on("popup") method',
        __filename,
        page,
        async (wrappedPage) => {
          await wrappedPage.setContent('<a href="about:blank" target="blank">CLICK HERE</a>');

          const [, newPage] = await Promise.all([
            wrappedPage.click('a'),
            new Promise<playwright.Page>((res) => {
              wrappedPage.on('popup', res);
            }),
          ]);
          await newPage.click('body');
        }
      );

      expect(beforeAllHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchSnapshot(
        'beforeAllHook'
      );
      expect(afterAllHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchSnapshot(
        'afterAllHook'
      );

      expect(beforeEachHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchSnapshot(
        'beforeEachHook'
      );

      expect(afterEachHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchSnapshot(
        'afterEachHook'
      );
    });

    test('page.once method', async () => {
      const {
        beforeAllHook,
        afterAllHook,
        beforeEachHook,
        afterEachHook,
      } = await runPageInstrumentorTest(
        'page.once method',
        __filename,
        page,
        async (wrappedPage) => {
          await wrappedPage.setContent('<a href="about:blank" target="blank">CLICK HERE</a>');

          const [, newPage] = await Promise.all([
            wrappedPage.click('a'),
            new Promise<playwright.Page>((res) => {
              wrappedPage.once('popup', res);
            }),
          ]);
          await newPage.click('body');
        }
      );

      expect(beforeAllHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchInlineSnapshot(`
        Array [
          Array [
            Object {
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "testContext": "cleaned",
            },
          ],
        ]
      `);
      expect(afterAllHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchInlineSnapshot(`
        Array [
          Array [
            Object {
              "endStatus": Object {
                "success": true,
              },
              "testContext": "cleaned",
            },
          ],
        ]
      `);

      expect(beforeEachHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchInlineSnapshot(`
        Array [
          Array [
            Object {
              "args": Array [
                "<a href=\\"about:blank\\" target=\\"blank\\">CLICK HERE</a>",
              ],
              "fnName": "setContent",
              "methodCallData": Array [
                Object {
                  "creationFunction": "setContent",
                  "pageId": 0,
                  "selector": undefined,
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
          Array [
            Object {
              "args": Array [
                "a",
              ],
              "fnName": "click",
              "methodCallData": Array [
                Object {
                  "creationFunction": "click",
                  "pageId": 0,
                  "selector": "a",
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
          Array [
            Object {
              "args": Array [
                "body",
              ],
              "fnName": "click",
              "methodCallData": Array [
                Object {
                  "creationFunction": "click",
                  "pageId": 1,
                  "selector": "body",
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
        ]
      `);

      expect(afterEachHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchInlineSnapshot(`
        Array [
          Array [
            Object {
              "args": Array [
                "<a href=\\"about:blank\\" target=\\"blank\\">CLICK HERE</a>",
              ],
              "fnName": "setContent",
              "instrumentedFunctionResult": "cleaned",
              "methodCallData": Array [
                Object {
                  "creationFunction": "setContent",
                  "pageId": 0,
                  "selector": undefined,
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
          Array [
            Object {
              "args": Array [
                "a",
              ],
              "fnName": "click",
              "instrumentedFunctionResult": "cleaned",
              "methodCallData": Array [
                Object {
                  "creationFunction": "click",
                  "pageId": 0,
                  "selector": "a",
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
          Array [
            Object {
              "args": Array [
                "body",
              ],
              "fnName": "click",
              "instrumentedFunctionResult": "cleaned",
              "methodCallData": Array [
                Object {
                  "creationFunction": "click",
                  "pageId": 1,
                  "selector": "body",
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
        ]
      `);
    });

    test('multiple pages playwright - waitForEvent', async () => {
      const {
        beforeAllHook,
        afterAllHook,
        beforeEachHook,
        afterEachHook,
      } = await runPageInstrumentorTest(
        'multiple pages playwright - waitForEvent',
        __filename,
        page,
        async (wrappedPage) => {
          await wrappedPage.setContent('<a href="about:blank" target="blank">CLICK HERE</a>');
          const [, newPage] = await Promise.all([
            wrappedPage.click('a'),
            wrappedPage.waitForEvent('popup'),
          ]);
          await newPage.click('body');
        }
      );

      expect(beforeAllHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchInlineSnapshot(`
        Array [
          Array [
            Object {
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "testContext": "cleaned",
            },
          ],
        ]
      `);
      expect(afterAllHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchInlineSnapshot(`
        Array [
          Array [
            Object {
              "endStatus": Object {
                "success": true,
              },
              "testContext": "cleaned",
            },
          ],
        ]
      `);

      expect(beforeEachHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchInlineSnapshot(`
        Array [
          Array [
            Object {
              "args": Array [
                "<a href=\\"about:blank\\" target=\\"blank\\">CLICK HERE</a>",
              ],
              "fnName": "setContent",
              "methodCallData": Array [
                Object {
                  "creationFunction": "setContent",
                  "pageId": 0,
                  "selector": undefined,
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
          Array [
            Object {
              "args": Array [
                "a",
              ],
              "fnName": "click",
              "methodCallData": Array [
                Object {
                  "creationFunction": "click",
                  "pageId": 0,
                  "selector": "a",
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
          Array [
            Object {
              "args": Array [
                "body",
              ],
              "fnName": "click",
              "methodCallData": Array [
                Object {
                  "creationFunction": "click",
                  "pageId": 1,
                  "selector": "body",
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
        ]
      `);

      expect(afterEachHook.mock.calls.map(cleanJestMockFunctionCalls)).toMatchInlineSnapshot(`
        Array [
          Array [
            Object {
              "args": Array [
                "<a href=\\"about:blank\\" target=\\"blank\\">CLICK HERE</a>",
              ],
              "fnName": "setContent",
              "instrumentedFunctionResult": "cleaned",
              "methodCallData": Array [
                Object {
                  "creationFunction": "setContent",
                  "pageId": 0,
                  "selector": undefined,
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
          Array [
            Object {
              "args": Array [
                "a",
              ],
              "fnName": "click",
              "instrumentedFunctionResult": "cleaned",
              "methodCallData": Array [
                Object {
                  "creationFunction": "click",
                  "pageId": 0,
                  "selector": "a",
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
          Array [
            Object {
              "args": Array [
                "body",
              ],
              "fnName": "click",
              "instrumentedFunctionResult": "cleaned",
              "methodCallData": Array [
                Object {
                  "creationFunction": "click",
                  "pageId": 1,
                  "selector": "body",
                  "text": undefined,
                },
              ],
              "proxyContext": "cleaned",
              "rootPage": "cleaned",
              "stepResult": "cleaned",
              "testContext": "cleaned",
            },
          ],
        ]
      `);
    });
  });
});
