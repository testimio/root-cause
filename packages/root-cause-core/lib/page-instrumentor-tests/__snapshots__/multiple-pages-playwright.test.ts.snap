// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiple pages Playwright chromium page.addListener("popup") method: afterAllHook 1`] = `
Array [
  Array [
    Object {
      "endStatus": Object {
        "success": true,
      },
      "testContext": "cleaned",
    },
  ],
]
`;

exports[`Multiple pages Playwright chromium page.addListener("popup") method: afterEachHook 1`] = `
Array [
  Array [
    Object {
      "args": Array [
        "<a href=\\"about:blank\\" target=\\"blank\\">CLICK HERE</a>",
      ],
      "fnName": "setContent",
      "instrumentedFunctionResult": "cleaned",
      "methodCallData": Array [
        Object {
          "creationFunction": "setContent",
          "pageId": 0,
          "selector": undefined,
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
  Array [
    Object {
      "args": Array [
        "a",
      ],
      "fnName": "click",
      "instrumentedFunctionResult": "cleaned",
      "methodCallData": Array [
        Object {
          "creationFunction": "click",
          "pageId": 0,
          "selector": "a",
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
  Array [
    Object {
      "args": Array [
        "body",
      ],
      "fnName": "click",
      "instrumentedFunctionResult": "cleaned",
      "methodCallData": Array [
        Object {
          "creationFunction": "click",
          "pageId": 1,
          "selector": "body",
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
]
`;

exports[`Multiple pages Playwright chromium page.addListener("popup") method: beforeAllHook 1`] = `
Array [
  Array [
    Object {
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "testContext": "cleaned",
    },
  ],
]
`;

exports[`Multiple pages Playwright chromium page.addListener("popup") method: beforeEachHook 1`] = `
Array [
  Array [
    Object {
      "args": Array [
        "<a href=\\"about:blank\\" target=\\"blank\\">CLICK HERE</a>",
      ],
      "fnName": "setContent",
      "methodCallData": Array [
        Object {
          "creationFunction": "setContent",
          "pageId": 0,
          "selector": undefined,
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
  Array [
    Object {
      "args": Array [
        "a",
      ],
      "fnName": "click",
      "methodCallData": Array [
        Object {
          "creationFunction": "click",
          "pageId": 0,
          "selector": "a",
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
  Array [
    Object {
      "args": Array [
        "body",
      ],
      "fnName": "click",
      "methodCallData": Array [
        Object {
          "creationFunction": "click",
          "pageId": 1,
          "selector": "body",
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
]
`;

exports[`Multiple pages Playwright chromium page.on("popup") method: afterAllHook 1`] = `
Array [
  Array [
    Object {
      "endStatus": Object {
        "success": true,
      },
      "testContext": "cleaned",
    },
  ],
]
`;

exports[`Multiple pages Playwright chromium page.on("popup") method: afterEachHook 1`] = `
Array [
  Array [
    Object {
      "args": Array [
        "<a href=\\"about:blank\\" target=\\"blank\\">CLICK HERE</a>",
      ],
      "fnName": "setContent",
      "instrumentedFunctionResult": "cleaned",
      "methodCallData": Array [
        Object {
          "creationFunction": "setContent",
          "pageId": 0,
          "selector": undefined,
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
  Array [
    Object {
      "args": Array [
        "a",
      ],
      "fnName": "click",
      "instrumentedFunctionResult": "cleaned",
      "methodCallData": Array [
        Object {
          "creationFunction": "click",
          "pageId": 0,
          "selector": "a",
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
  Array [
    Object {
      "args": Array [
        "body",
      ],
      "fnName": "click",
      "instrumentedFunctionResult": "cleaned",
      "methodCallData": Array [
        Object {
          "creationFunction": "click",
          "pageId": 1,
          "selector": "body",
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
]
`;

exports[`Multiple pages Playwright chromium page.on("popup") method: beforeAllHook 1`] = `
Array [
  Array [
    Object {
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "testContext": "cleaned",
    },
  ],
]
`;

exports[`Multiple pages Playwright chromium page.on("popup") method: beforeEachHook 1`] = `
Array [
  Array [
    Object {
      "args": Array [
        "<a href=\\"about:blank\\" target=\\"blank\\">CLICK HERE</a>",
      ],
      "fnName": "setContent",
      "methodCallData": Array [
        Object {
          "creationFunction": "setContent",
          "pageId": 0,
          "selector": undefined,
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
  Array [
    Object {
      "args": Array [
        "a",
      ],
      "fnName": "click",
      "methodCallData": Array [
        Object {
          "creationFunction": "click",
          "pageId": 0,
          "selector": "a",
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
  Array [
    Object {
      "args": Array [
        "body",
      ],
      "fnName": "click",
      "methodCallData": Array [
        Object {
          "creationFunction": "click",
          "pageId": 1,
          "selector": "body",
          "text": undefined,
        },
      ],
      "proxyContext": "cleaned",
      "rootPage": "cleaned",
      "stepResult": "cleaned",
      "testContext": "cleaned",
    },
  ],
]
`;
